lib/ast/AST.cpp

```diff
@@ -928,6 +928,106 @@ sptr<KOREPattern> KORECompositePattern::filterSubstitution(PrettyPrintData const
   }
 }
 
+sptr<KOREPattern> KORECompositePattern::expandMacros(SubsortMap const& subsorts, SymbolMap const& overloads, std::vector<ptr<KOREDeclaration>> const& macros, bool reverse, std::set<size_t> &appliedRules) {
+  size_t i = 0;
+  for (auto &decl : macros) {
+    if ((decl->getAttributes().count("macro") || decl->getAttributes().count("macro-rec")) && reverse) {
+      i++;
+      continue;
+    }
+    auto axiom = dynamic_cast<KOREAxiomDeclaration *>(decl.get());
+    auto equals = dynamic_cast<KORECompositePattern *>(axiom->getPattern().get());
+    auto lhs = equals->arguments[reverse ? 1 : 0];
+    auto rhs = equals->arguments[reverse ? 0 : 1];
+    substitution subst;
+    bool matches = lhs->matches(subst, subsorts, overloads, shared_from_this());
+    if (matches && (decl->getAttributes().count("macro-rec") || decl->getAttributes().count("alias-rec") || !appliedRules.count(i))) {
+      std::set<size_t> oldAppliedRules = appliedRules;
+      appliedRules.insert(i);
+      auto result = rhs->substitute(subst)->expandMacros(subsorts, overloads, macros, reverse, appliedRules);
+      appliedRules = oldAppliedRules;
+      return result;
+    }
+    i++;
+  }
+  if (arguments.empty()) {
+    return shared_from_this();
+  }
+  sptr<KORECompositePattern> result = KORECompositePattern::Create(constructor.get());
+  for (auto &arg : arguments) {
+    result->addArgument(arg->expandMacros(subsorts, overloads, macros, reverse));
+  }
+  return result;
+}
+
+bool KORECompositePattern::matches(substitution &subst, SubsortMap const& subsorts, SymbolMap const& overloads, sptr<KOREPattern> subject) {
+  auto subj = dynamic_cast<KORECompositePattern *>(subject.get());
+  if (!subj) {
+    return false;
+  }
+  if (*subj->getConstructor() != *getConstructor()) {
+    if (subj->getConstructor()->getName() == "inj" && getConstructor()->getName() == "inj") {
+      if (*subj->getConstructor()->getFormalArguments()[1] != *getConstructor()->getFormalArguments()[1]) {
+        return false;
+      }
+      sptr<KORESort> a = subj->getConstructor()->getFormalArguments()[0];
+      sptr<KORESort> b = getConstructor()->getFormalArguments()[0];
+      sptr<KORECompositePattern> ab = KORECompositePattern::Create("inj");
+      ab->getConstructor()->addArgument(a);
+      ab->getConstructor()->addArgument(b);
+      ab->addArgument(subj->getArguments()[0]);
+      if (subsorts.count(b.get()) && subsorts.at(b.get()).count(a.get())) {
+        //pass
```

Comment:
Why does this case "fall through" and get resolved as argument matching below? I.e., why doesn't this just return `false`, or why doesn't `inj` get decomposed as in the other subsorting case?

Is it because the only case where it would match would be when `this` is an `inj(variable:b)`, and it's ok to have a substitution from `variable` to `stuff:a` without having an inj on top of it, i.e. inj(a, b)(stuff)?

Reply:
This code was based originally on the code for pattern matching in the llvm backend, and originally this case worked the way it did because of the way that algorithm worked, but I think I mistranslated it to the new domain. I'm going to go ahead and do what you suggest and decompose it via the subsort.